04-onion-routing.md

```diff
@@ -237,13 +270,48 @@ Field descriptions:
      `outgoing_cltv_value`, whether it is the final node or not, to avoid
      leaking its position in the route.
 
-   * `padding`: This field is for future use and also for ensuring that future non-0-`realm`
-     `per_hop`s won't change the overall `hops_data` size.
+   * `padding`: This field is for future use.
 
 When forwarding HTLCs, nodes MUST construct the outgoing HTLC as specified within
-`per_hop` above; otherwise, deviation from the specified HTLC parameters
+`legacy_per_hop` above; otherwise, deviation from the specified HTLC parameters
 may lead to extraneous routing failure.
 
+## `tlv_hop_data` Payload Format
+
+This is a more flexible format, which avoids the redundant
+`short_channel_id` field for the final node.  The range of `realm`
+values also allows use of more than one frame.
+
+1. tlv: `tlv_hop_data`
+2. types:
```

Comment:
Rather than unroll these fields and take up 4 types, we can instead have them all packed under a _single_ type. All the fields are fixed length, so there's no additional signalling overhead and we also save a few bytes as well. When allocating space in the TLV namespace we should prefer to pack several types into one, if all the types are related to a discrete use case. This model promotes more efficient utilization of the payload space.

Reply:
Well like we discussed in the IRC meeting some of the types would be omitted, for example in the last `hop_payload`. So while I agree that we may avoid defining types that must always be used in combination, we should also keep in mind that swapping out or omitting individual types is also an important feature here.